File: src/bidi-connection.ts
/**
 * WebDriver BiDi WebSocket Connection
 *
 * Handles direct WebSocket connection to Firefox's BiDi protocol endpoint.
 */

import WebSocket from 'ws';
import { log, logDebug } from './logger.js';
import type { BiDiConnectionOptions, PendingCommand, EventHandler } from './types.js';

export class BiDiConnection {
  private ws: WebSocket | null = null;
  private nextId: number = 1;
  private pendingCommands: Map<number, PendingCommand> = new Map();
  private eventHandlers: Set<EventHandler> = new Set();
  private namedEventHandlers: Map<string, Set<(params: any) => void>> = new Map();
  private connected: boolean = false;
  private options: BiDiConnectionOptions;
  private _directSend: ((msg: object) => void) | null = null;

  constructor(options: BiDiConnectionOptions = {}) {
    this.options = {
      connectionTimeout: options.connectionTimeout || 5000,
      commandTimeout: options.commandTimeout || 20000,
    };
  }

  /**
   * Wire the connection directly to an in-process BiDi server (no WebSocket).
   * sendFn receives outgoing BiDi commands as parsed objects.
   * Call receiveDirect() to inject incoming messages from the server.
   */
  connectDirect(sendFn: (msg: object) => void): void {
    this._directSend = sendFn;
    this.connected = true;
  }

  receiveDirect(message: object): void {
    this.handleMessage(message);
  }

  /**
   * Connect to a browser BiDi WebSocket endpoint.
   * Pass a port number (Firefox) or a full WebSocket URL (Chrome).
   */
  async connect(portOrUrl: number | string): Promise<void> {
    logDebug(`Attempting to connect to BiDi on ${portOrUrl}`);

    // Chrome provides a full URL; Firefox just needs the port
    const endpoints = typeof portOrUrl === 'string'
      ? [portOrUrl]
      : [
          `ws://127.0.0.1:${portOrUrl}`,
          `ws://127.0.0.1:${portOrUrl}/session`,
        ];

    let lastError: Error | null = null;

    for (const endpoint of endpoints) {
      try {
        await this.connectToEndpoint(endpoint);
        log(`✅ Connected to BiDi at ${endpoint}`);
        this.connected = true;
        return;
      } catch (error) {
        lastError = error as Error;
        logDebug(`Failed to connect to ${endpoint}: ${lastError.message}`);
      }
    }

    // All endpoints failed
    throw new Error(
      `Failed to connect to browser BiDi at ${portOrUrl}.\n` +
      `Last error: ${lastError?.message || 'Unknown error'}`
    );
  }

  /**
   * Try to connect to a specific WebSocket endpoint
   */
  private async connectToEndpoint(endpoint: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(endpoint);
      let timeoutId: NodeJS.Timeout;

      const cleanup = () => {
        clearTimeout(timeoutId);
        ws.removeAllListeners();
      };

      timeoutId = setTimeout(() => {
        cleanup();
        ws.close();
        reject(new Error('Connection timeout'));
      }, this.options.connectionTimeout);

      ws.on('open', () => {
        cleanup();
        this.ws = ws;
        this.setupWebSocketHandlers();
        resolve();
      });

      ws.on('error', (error) => {
        cleanup();
        logDebug(`WebSocket error for ${endpoint}: ${error.message || String(error)}`);
        reject(error);
      });
    });
  }

  /**
   * Set up WebSocket event handlers
   */
  private setupWebSocketHandlers(): void {
    if (!this.ws) return;

    this.ws.on('message', (data: WebSocket.Data) => {
      try {
        const message = JSON.parse(data.toString());
        this.handleMessage(message);
      } catch (error) {
        logDebug(`Failed to parse BiDi message: ${error}`);
      }
    });

    this.ws.on('close', () => {
      logDebug('BiDi WebSocket closed');
      this.connected = false;
      this.ws = null;

      // Reject all pending commands
      for (const [id, pending] of this.pendingCommands.entries()) {
        clearTimeout(pending.timeout);
        pending.reject(new Error('WebSocket closed'));
      }
      this.pendingCommands.clear();
    });

    this.ws.on('error', (error) => {
      logDebug(`BiDi WebSocket error: ${error.message}`);
    });
  }

  /**
   * Handle incoming WebSocket message
   */
  private handleMessage(message: any): void {
    // Command response (has 'id' field)
    if (message.id !== undefined) {
      const pending = this.pendingCommands.get(message.id);
      if (pending) {
        clearTimeout(pending.timeout);
        this.pendingCommands.delete(message.id);

        if (message.error) {
          pending.reject(
            new Error(`BiDi error in ${pending.method}: ${JSON.stringify(message.error)}`)
          );
        } else {
          logDebug(`← BiDi result: ${pending.method} ${JSON.stringify(message.result || {})}`);
          pending.resolve(message.result || {});
        }
      }
      return;
    }

    // Event (no 'id' field, has 'method' field)
    if (message.method) {
      logDebug(`BiDi event: ${message.method}`);
      for (const handler of this.eventHandlers) {
        try {
          handler(message);
        } catch (error) {
          logDebug(`Error in event handler: ${error}`);
        }
      }
      const named = this.namedEventHandlers.get(message.method);
      if (named) {
        for (const handler of named) {
          try {
            handler(message.params);
          } catch (error) {
            logDebug(`Error in named event handler for ${message.method}: ${error}`);
          }
        }
      }
    }
  }

  /**
   * Send BiDi command and wait for response
   */
  async sendCommand(method: string, params: any = {}): Promise<any> {
    if (!this._directSend) {
      if (!this.ws) {
        throw new Error('Not connected to browser BiDi');
      }
      if (this.ws.readyState !== WebSocket.OPEN) {
        throw new Error('Not connected to browser BiDi');
      }
    }

    const id = this.nextId++;
    const command = { id, method, params };

    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        this.pendingCommands.delete(id);
        reject(new Error(`BiDi command timeout: ${method}`));
      }, this.options.commandTimeout);

      this.pendingCommands.set(id, { resolve, reject, timeout: timeoutId, method });

      if (this._directSend) {
        logDebug(`→ BiDi command (direct): ${method} ${JSON.stringify(params)}`);
        this._directSend(command);
      } else {
        logDebug(`→ BiDi command: ${method} ${JSON.stringify(params)}`);
        this.ws!.send(JSON.stringify(command));
      }
    });
  }

  /**
   * Subscribe to BiDi events
   */
  async subscribe(events: string[], contexts?: string[]): Promise<void> {
    const params: any = { events };
    if (contexts) {
      params.contexts = contexts;
    }

    await this.sendCommand('session.subscribe', params);
    logDebug(`Subscribed to events: ${events.join(', ')}`);
  }

  /**
   * Register a handler for raw WebSocket-level events ('message', 'close', 'error').
   * For BiDi protocol events use onEvent() instead.
   */
  on(event: 'message' | 'close' | 'error', handler: EventHandler): void {
    if (event === 'message') {
      this.eventHandlers.add(handler);
    } else if (this.ws) {
      this.ws.on(event, handler);
    }
  }

  /**
   * Remove a raw event handler registered with on().
   */
  off(event: 'message', handler: EventHandler): void {
    if (event === 'message') {
      this.eventHandlers.delete(handler);
    }
  }

  /**
   * Register a handler for a specific BiDi event method (e.g. 'log.entryAdded').
   * The handler receives event.params directly.
   * Remember to call subscribe() for the event type before expecting events.
   */
  onEvent(method: string, handler: (params: any) => void): void {
    if (!this.namedEventHandlers.has(method)) {
      this.namedEventHandlers.set(method, new Set());
    }
    this.namedEventHandlers.get(method)!.add(handler);
  }

  /**
   * Remove a handler registered with onEvent().
   */
  offEvent(method: string, handler: (params: any) => void): void {
    this.namedEventHandlers.get(method)?.delete(handler);
  }

  /**
   * Get underlying WebSocket (for compatibility)
   */
  get socket(): WebSocket | null {
    return this.ws;
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    if (this._directSend) return this.connected;
    return this.connected && this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }

  /**
   * Close connection
   */
  async close(): Promise<void> {
    if (this._directSend) {
      logDebug('Closing BiDi direct connection');
      for (const [, pending] of this.pendingCommands.entries()) {
        clearTimeout(pending.timeout);
        pending.reject(new Error('Connection closing'));
      }
      this.pendingCommands.clear();
      this.eventHandlers.clear();
      this.namedEventHandlers.clear();
      this._directSend = null;
      this.connected = false;
      return;
    }

    if (this.ws) {
      logDebug('Closing BiDi WebSocket connection');

      // Clear all pending commands
      for (const [, pending] of this.pendingCommands.entries()) {
        clearTimeout(pending.timeout);
        pending.reject(new Error('Connection closing'));
      }
      this.pendingCommands.clear();

      // Clear event handlers
      this.eventHandlers.clear();
      this.namedEventHandlers.clear();

      // Close WebSocket
      if (this.ws.readyState === WebSocket.OPEN || this.ws.readyState === WebSocket.CONNECTING) {
        this.ws.close();
      }

      this.ws = null;
      this.connected = false;
    }
  }
}

File: python/firefox_bidi_client/bidi_connection.py
import asyncio
import json
from typing import Any, Callable, Dict, Optional, Set

import websockets

from .logger import log, log_debug


class BiDiConnection:
    def __init__(self, connection_timeout: float = 5.0, command_timeout: float = 10.0):
        self._ws = None
        self._next_id: int = 1
        self._pending: Dict[int, asyncio.Future] = {}
        self._pending_methods: Dict[int, str] = {}
        self._event_handlers: Set[Callable] = set()
        self._named_event_handlers: Dict[str, Set[Callable]] = {}
        self._connected: bool = False
        self._connection_timeout = connection_timeout
        self._command_timeout = command_timeout
        self._receive_task: Optional[asyncio.Task] = None

    async def connect(self, port: int) -> None:
        log_debug(f"Attempting to connect to BiDi on port {port}")
        endpoints = [
            f"ws://127.0.0.1:{port}",
            f"ws://127.0.0.1:{port}/session",
        ]
        last_error: Optional[Exception] = None
        for endpoint in endpoints:
            try:
                await self._connect_to_endpoint(endpoint)
                log(f"Connected to Firefox BiDi at {endpoint}")
                self._connected = True
                return
            except Exception as e:
                last_error = e
                log_debug(f"Failed to connect to {endpoint}: {e}")

        raise ConnectionError(
            f"Failed to connect to Firefox on port {port}.\n"
            f"Last error: {last_error}"
        )

    async def _connect_to_endpoint(self, endpoint: str) -> None:
        ws = await asyncio.wait_for(
            websockets.connect(endpoint),
            timeout=self._connection_timeout,
        )
        self._ws = ws
        self._receive_task = asyncio.create_task(self._receive_loop())

    async def _receive_loop(self) -> None:
        try:
            async for message in self._ws:
                try:
                    self._handle_message(json.loads(message))
                except Exception as e:
                    log_debug(f"Failed to parse BiDi message: {e}")
        except Exception as e:
            log_debug(f"BiDi receive loop ended: {e}")
        finally:
            self._connected = False
            for fut in self._pending.values():
                if not fut.done():
                    fut.set_exception(ConnectionError("WebSocket closed"))
            self._pending.clear()

    def _handle_message(self, message: Any) -> None:
        if message.get("id") is not None:
            msg_id = message["id"]
            fut = self._pending.pop(msg_id, None)
            method = self._pending_methods.pop(msg_id, "unknown")
            if fut and not fut.done():
                if "error" in message:
                    fut.set_exception(
                        RuntimeError(f"BiDi error in {method}: {json.dumps(message['error'])}")
                    )
                else:
                    fut.set_result(message.get("result", {}))
            return

        if "method" in message:
            log_debug(f"BiDi event: {message['method']}")
            for handler in list(self._event_handlers):
                try:
                    handler(message)
                except Exception as e:
                    log_debug(f"Error in event handler: {e}")
            named = self._named_event_handlers.get(message["method"])
            if named:
                for handler in list(named):
                    try:
                        handler(message.get("params", {}))
                    except Exception as e:
                        log_debug(f"Error in named event handler for {message['method']}: {e}")

    async def send_command(self, method: str, params: Any = None) -> Any:
        if self._ws is None:
            raise ConnectionError("Not connected to Firefox BiDi")

        if params is None:
            params = {}

        cmd_id = self._next_id
        self._next_id += 1

        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[cmd_id] = fut
        self._pending_methods[cmd_id] = method

        log_debug(f"→ BiDi command: {method}")
        await self._ws.send(json.dumps({"id": cmd_id, "method": method, "params": params}))

        try:
            return await asyncio.wait_for(asyncio.shield(fut), timeout=self._command_timeout)
        except asyncio.TimeoutError:
            self._pending.pop(cmd_id, None)
            self._pending_methods.pop(cmd_id, None)
            raise TimeoutError(f"BiDi command timeout: {method}")

    async def subscribe(self, events: list, contexts: Optional[list] = None) -> None:
        params: dict = {"events": events}
        if contexts:
            params["contexts"] = contexts
        await self.send_command("session.subscribe", params)
        log_debug(f"Subscribed to events: {', '.join(events)}")

    def on_event(self, method: str, handler: Callable) -> None:
        """Register a handler for a specific BiDi event (e.g. 'log.entryAdded').
        The handler receives event.params directly. Call subscribe() first."""
        if method not in self._named_event_handlers:
            self._named_event_handlers[method] = set()
        self._named_event_handlers[method].add(handler)

    def off_event(self, method: str, handler: Callable) -> None:
        if method in self._named_event_handlers:
            self._named_event_handlers[method].discard(handler)

    def on(self, handler: Callable) -> None:
        """Register a handler for all raw BiDi messages."""
        self._event_handlers.add(handler)

    def off(self, handler: Callable) -> None:
        self._event_handlers.discard(handler)

    def is_connected(self) -> bool:
        return self._connected and self._ws is not None

    async def close(self) -> None:
        if self._receive_task:
            self._receive_task.cancel()
            try:
                await self._receive_task
            except asyncio.CancelledError:
                pass
            self._receive_task = None

        if self._ws:
            log_debug("Closing BiDi WebSocket connection")
            for fut in self._pending.values():
                if not fut.done():
                    fut.set_exception(ConnectionError("Connection closing"))
            self._pending.clear()
            self._event_handlers.clear()
            self._named_event_handlers.clear()
            await self._ws.close()
            self._ws = None
            self._connected = False
